File: htdocs/register/paypal.php
<?php

/*
 * This software is licensed under the GPLv3 license, included as
 * ./GPLv3-LICENSE.txt in the source distribution.
 *
 * Portions created by Brett Wilson are Copyright 2008 Brett Wilson.
 * All rights reserved.
 */



function process_ipn()
{
	$_POST['cmd'] = "_notify-validate";

	$curl = curl_init();
	curl_setopt($curl, CURLOPT_URL, "https://www.paypal.com/cgi-bin/webscr");
	curl_setopt($curl, CURLOPT_POST, 1);
	curl_setopt($curl, CURLOPT_POSTFIELDS, $_POST);
	curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
	$response = curl_exec($curl);
	$error = curl_error($curl);
	curl_close ($curl);

	if ($response == 'VERIFIED')
	{
		## txid is a unique key, if a previous exists, this one will overwrite
		## the only time this would concievably happen is when an echeck clears

		$inst = getps("replace into payments (txid,type,date,status,driverid,eventid,amount) values (?,?,?,?,?,?,?)");

  		list($eventid, $driverid) = explode('.', $_POST['custom']);
		$args = array($_POST['txn_id'], $_POST['payment_type'], (time()*1000),
				$_POST['payment_status'], $driverid, $eventid, $_POST['payment_gross']);
		$inst->execute($args);
	}

	return;
}

?>

File: web/nwrsc/lib/paypal.py
from pylons import request
from nwrsc.model import *

import logging
log = logging.getLogger(__name__)

class PayPalIPN(object):

	def ipn(self):
		data = dict(request.POST.items())
		args = { 'cmd': '_notify-validate' }
		args.update(dict(request.POST.items()))
		log.debug("Paypal IPN sends: %s" % args)

		result = urllib.urlopen("https://www.paypal.com/cgi-bin/webscr", urllib.urlencode(args)).read().strip()
		if result == 'VERIFIED':
			## txid is a unique key, if a previous exists, this one will overwrite
			## the only time this would concievably happen is when an echeck clears
			log.info("Paypal IPN: %s" % result)
			tx = self.session.query(Payment).get(data.get('txn_id'))
			if tx is None:
				tx = Payment()
				self.session.add(tx)
				tx.txid = data.get('txn_id')
			tx.type = data.get('payment_type')
			tx.date = datetime.datetime.now()
			tx.status = data.get('payment_status')
			tx.amount = data.get('mc_gross')
			parts = map(int, data.get('custom').split('.'))
			if len(parts) != 2:
				log.error("Paypal IPN: invalid custom: %s" % data.get('custom'))
			else:
				tx.eventid = parts[0]
				tx.driverid = parts[1]
				self.session.commit()
		else:
			log.error("Paypal result not verified: '%s'", result)

